### PR TITLE
fix: ship unshipped stack review comments (#3–#15 via #17)

### PR DESCRIPTION
**Pull request by arthrod**
## Summary

PR #17 merged with **zero implementable review comments** (CodeRabbit skipped: 285 files over the 100-file limit; Sourcery/Gemini/Qodo unavailable). The Graphite stack PRs #3–#15 that landed inside #17 still had **21 unshipped inline comments**.

This PR gates every one and implements the survivors on `main`.

### Ledger (21 inline + 6 #17 meta)

| Verdict | Count | Notes |
|---------|------:|-------|
| implement | 20 | correctness + tests + nits |
| push-back | 1 | #7 subprocess “dangerous” — fixed `["soffice", "--version"]` argv, no external input |
| #17 meta n/a | 6 | bot skip/sunset/rate-limit only |

### Highlights

- **export-results-md**: prefer `holdout_mode`-stamped lines over pre-holdout legacy; ITT median/mean in rerun selection; fix `""""` docstring
- **coverage_matrix**: skip non-numeric scores; lowercase stem join for median tables
- **noise_floor**: refuse non-finite/negative sigma (would disable the gate)
- **canary**: same soffice resolution as renderer; stdout+stderr version; `invalid-spec` on corrupt JSON
- **word-validate / scoreboard**: line-anchored WV parse; fail closed on modal probe `None` (exit 2); JSON keys by path not stem
- **tests**: pagefair multi/both-unmatched, config non-list `extra_oracle_dirs`, null_score pin, holdout-prefer export, etc.

### Test plan

- [x] `uv run pytest` targeted suites (121 passed)
- [x] `bunx vitest run scripts/redline_scoreboard.test.ts scripts/update-readme-ranking.test.ts` (26 passed)

## Summary by Sourcery

Harden renderer canary, Word validation, results export, and coverage reporting while aligning CLI behavior and tests with real-world edge cases.

New Features:
- Treat malformed or unreadable renderer canary specs as an explicit invalid-spec status surfaced in both the CLI gate and dedicated canary command.
- Prefer holdout-aware benchmark rows and ITT-based quality metrics when selecting headline results from bench.jsonl, and propagate holdout_mode through exported rows.

Bug Fixes:
- Resolve soffice using the same logic as the renderer and parse its version from combined stdout and stderr to support more LibreOffice builds.
- Normalize score key casing when joining coverage mappings to vendor scores so mixed-case stems still contribute to median tables and unjoined key reporting.
- Anchor word-validate scoreboard parsing to per-line outcome prefixes so filenames and error text containing VALID/INVALID tokens do not corrupt counts.
- Ensure Word modal probe failures cause word-validate to fail closed instead of silently treating the probe as no-modal, avoiding mislabelled UNJUDGEABLE outcomes.
- Prevent noise-floor epsilon loading from accepting non-finite or negative sigma values that could disable or invert the gate.
- Allow the oracle manifest CLI to treat corrupted manifests as drift rather than silently appearing clean.
- Correct README ranking script main-guard path comparisons so direct and platform-specific invocations are reliably detected.

Enhancements:
- Key word-validate JSON records by path-relative document identifiers instead of bare stems to disambiguate files and better track per-target results.
- Clarify CLI help around oracle manifest checks to distinguish drift failures from missing-manifest warnings for synthetic corpora.
- Slightly simplify internal noise-floor and CLI utility code while keeping behavior intact.
- Add null_score assertions to score_v2 tests to pin expected alignment with overall_score for null baselines.

Tests:
- Extend export-results, coverage matrix, pagefair, canary, config, oracle manifest, noise-floor, Word validation scoreboard, and scoring tests to cover new edge cases and guard against regressions.

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved document validation so modal-probe failures are reported as errors instead of being misclassified.
  - Corrected validation result counting to ignore status words in filenames and messages.
  - Improved handling of malformed renderer specifications with clearer diagnostics.
  - Fixed case-insensitive score matching and invalid noise-floor value handling.
  - Improved compatibility when launching ranking tools on Windows and with relative paths.

- **Improvements**
  - Ranking now prioritizes holdout-aware results and uses additional quality criteria for tie-breaking.
  - Word-validation results now use paths relative to the selected target directory when possible.

### CODE REVIEW

**Comment by arthrod** [comment]
@coderabbitai review

File: src/neurotic_docx_bench/coverage_matrix.py [hunk 1]
```diff
@@ -406,11 +414,19 @@ def render_markdown(
         for stem, info in coverage["pairs"].items():
             for tag in info["features"] + info["revisions"]:
                 stems_by_tag.setdefault(tag, []).append(stem)
+        # Score keys are pipeline-canonicalized (lowercase); stems keep mapping casing.
+        for vendor in vendors:
+            scores = scores_by_vendor[vendor]
```
[the remark this answers is not in the document]
Comment (arthrod) [reply src/neurotic_docx_bench/coverage_matrix.py]: Fixed in b2558111. `render_markdown` builds a local `lower_scores_by_vendor` and never mutates the caller dict. Test asserts `MixedCase` key preserved.

File: tests/test_coverage_matrix.py [hunk 2]
```diff
@@ -346,6 +346,21 @@ def test_render_markdown_vendor_medians(tmp_path: Path) -> None:
...
+def test_render_markdown_joins_mixed_case_stems_to_lowercase_scores(tmp_path: Path) -> None:
+    """Mapping stems keep casing; bench score keys are lowercased — join must normalize."""
+    root, src, red = _corpus(tmp_path)
+    make_docx(red / "File_A_File_B_redline.docx", _INS)
+    mapping = write_mapping(root / "map.csv", [{
+        "pair_stem": "File_A_File_B",
+        "docx_source_base": "a.docx", "docx_source_next": "b.docx",
+        "redline_docx": "File_A_File_B_redline.docx",
+    }])
+    coverage = build_coverage([mapping], [src], [red])
```
[the remark this answers is not in the document]
Comment (arthrod) [reply tests/test_coverage_matrix.py:359]: Fixed in b2558111. Tests: `test_latest_scores_by_vendor_filters_non_numeric_and_non_finite`, `test_latest_scores_by_vendor_empty_clears_stale`, `test_unjoined_score_keys_ignores_case_differences`.

File: tests/test_export_results_md.py [hunk 3]
```diff
@@ -549,3 +549,24 @@ def test_holdout_blurb_avoids_a_size_claim_when_vendors_disagree(tmp_path: Path)
...
+def test_rows_prefer_holdout_aware_over_legacy_full_corpus(tmp_path: Path) -> None:
+    """Once a holdout_mode=excluded line exists, pre-holdout lines (missing the field)
+    must not win headline tables just because they scored more docs."""
+    p = tmp_path / "bench.jsonl"
+    legacy = _fidelity_line("v", corpus_revision="rev", mean=99.0)
+    legacy["n_docs"] = 803
+    legacy["tool_version"] = "1.0"
+    legacy["timestamp"] = "2026-07-01T00:00:00+00:00"
+    # no holdout_mode field
+    modern = _fidelity_line("v", corpus_revision="rev", mean=70.0)
```
[the remark this answers is not in the document]
Comment (arthrod) [reply tests/test_export_results_md.py:567]: Fixed in b2558111. Test: `test_rows_prefer_itt_over_completed_only_when_both_holdout_aware` (equal timestamps → ITT quality wins).

File: scripts/update-readme-ranking.ts [hunk 4]
```diff
@@ -25,7 +25,8 @@
  * pooling used by export-results-md (best by n, then lower median).
  */
 import { existsSync, readFileSync, readdirSync, writeFileSync } from "node:fs";
-import { join, resolve } from "node:path";
+import { basename, join, resolve } from "node:path";
+import { fileURLToPath } from "node:url";
```
[the remark this answers is not in the document]
Comment (arthrod) [reply scripts/update-readme-ranking.ts]: Fixed in b2558111. `isMainPath` is exact path or realpath only — basename fallback removed. Test: `scripts/update-readme-ranking.test.ts` same-basename different directory → false.

File: src/neurotic_docx_bench/canary.py [hunk 5]
```diff
@@ -74,18 +78,29 @@ def render_canary_hash(docx: Path, work_dir: Path, *, dpi: int = 144) -> str:
...
         data = json.loads(path.read_text())
-    except (json.JSONDecodeError, OSError):
-        return None
-    return data if isinstance(data, dict) and "docx" in data else None
+    except json.JSONDecodeError as exc:
+        raise CanarySpecError(f"malformed canary spec at {path}: {exc}") from exc
+    except OSError as exc:
+        raise CanarySpecError(f"unreadable canary spec at {path}: {exc}") from exc
+    if not isinstance(data, dict) or "docx" not in data:
+        raise CanarySpecError(f"canary spec at {path} missing required 'docx' field")
```
[the remark this answers is not in the document]
Comment (arthrod) [reply src/neurotic_docx_bench/canary.py]: Fixed in b2558111. `UnicodeDecodeError` → `CanarySpecError` / `invalid-spec`. Test: `test_check_non_utf8_spec_is_invalid_spec`.

File: src/neurotic_docx_bench/cli.py [hunk 6]
```diff
@@ -483,11 +483,20 @@ def word_validate(
 
     n_valid = n_invalid = n_unjudgeable = 0
     records: dict[str, dict[str, object]] = {}
+    target_resolved = target.resolve()
```
[the remark this answers is not in the document]
Comment (arthrod) [reply src/neurotic_docx_bench/cli.py:486]: Fixed in b2558111. Single-file target uses `target.parent` as record root so the key is the filename.

File: src/neurotic_docx_bench/coverage_matrix.py [hunk 7]
```diff
@@ -341,7 +341,14 @@ def latest_scores_by_vendor(jsonl_path: Path) -> dict[str, dict[str, float]]:
...
             if isinstance(vendor, str) and isinstance(scores, dict):
-                by_vendor[vendor] = {k: float(v) for k, v in scores.items()}
+                clean: dict[str, float] = {}
+                for k, v in scores.items():
+                    try:
+                        clean[str(k)] = float(v)
+                    except (TypeError, ValueError):
+                        continue
+                if clean:
+                    by_vendor[vendor] = clean
```
[the remark this answers is not in the document]
Comment (arthrod) [reply src/neurotic_docx_bench/coverage_matrix.py]: Fixed in b2558111. `math.isfinite` after float conversion; always assign (empty clears stale). Tests cover Infinity/NaN and clear-stale.

File: src/neurotic_docx_bench/render/word.py [hunk 8]
```diff
@@ -243,12 +247,22 @@ def validate_one(
...
             )
+        if modal is None:
+            # Probe unavailable (permissions / osascript failure): do not treat as
+            # "no modal" — that would mis-label a blocked repair dialog as UNJUDGEABLE.
+            _close_active_document()
+            proc.kill()
+            raise ModalProbeError(
+                "Word modal probe failed (grant Accessibility / System Events "
+                "permissions for osascript); cannot judge validity safely",
+            )
```
[the remark this answers is not in the document]
Comment (arthrod) [reply src/neurotic_docx_bench/render/word.py:267]: Fixed in b2558111. `proc.communicate()` after every `proc.kill()` (modal / probe-fail / budget).

File: tests/test_export_results_md.py [hunk 9]
```diff
@@ -549,3 +549,24 @@ def test_holdout_blurb_avoids_a_size_claim_when_vendors_disagree(tmp_path: Path)
...
+    modern["n_docs"] = 763
+    modern["tool_version"] = "1.0"
+    modern["holdout_mode"] = "excluded"
+    modern["timestamp"] = "2026-08-01T00:00:00+00:00"
+    p.write_text(json.dumps(legacy) + "\n" + json.dumps(modern) + "\n")
+    rows = exp.rows_from_jsonl(p)
+    assert len(rows) == 1
+    assert rows[0]["mean"] == 70.0
+    assert rows[0]["holdout_mode"] == "excluded"
```
[the remark this answers is not in the document]
Comment (arthrod) [reply tests/test_export_results_md.py:576]: Fixed in b2558111. Legacy row is now newer than modern so the test isolates holdout_aware precedence.

**Comment by arthrod** [comment]
## Ledger (this round)

| source | total | implement | push-back | n/a meta |
|--------|------:|----------:|----------:|---------:|
| Sourcery inline | 3 | 3 | 0 | 0 |
| CodeRabbit inline | 6 | 6 | 0 | 0 |
| Issue/review meta | 4 | 0 | 0 | 4 (walkthrough / already-reviewed / request) |
| **sum** | **13** | **9** | **0** | **4** |

`implemented + pushed-back + n/a == 13`.

Shipped in `b2558111`. Also re-ran:
```
python3 scripts/export-results-md.py
bun run update-readme-ranking
```
→ RESULTS.md / docs/RESULTS.md / README rankings regenerated (no table drift vs prior pin set).

@coderabbitai review
